op-e2e/opgeth/op_geth_test.go

```diff
@@ -1143,6 +1146,173 @@ func TestSoulGasToken(t *testing.T) {
+		})
+		_, err = opGeth.AddL2Block(ctx, transferEthTx, mintTx)
+		require.NoError(t, err)
+
+		receipt, err := opGeth.L2Client.TransactionReceipt(ctx, mintTx.Hash())
+		require.NoError(t, err)
+		assert.Equal(t, types.ReceiptStatusSuccessful, receipt.Status)
+
+		signer := types.LatestSigner(opGeth.L2ChainConfig)
+
+		tx := types.MustSignNewTx(targetPK, signer, &types.DynamicFeeTx{
+			ChainID:   big.NewInt(int64(cfg.DeployConfig.L2ChainID)),
+			Nonce:     1,
+			GasTipCap: big.NewInt(100),
+			GasFeeCap: big.NewInt(100000),
+			Gas:       1_000_001,
+			To:        &cfg.Secrets.Addresses().Alice,
+			Value:     big.NewInt(params.Ether), // this tx.value will lead to tx failure
+			Data:      nil,
+		})
+		_, err = opGeth.AddL2Block(ctx, tx)
+		// err is like: Invalid payload attributes:failed to force-include tx: ..., err: insufficient funds for gas * price + value
+		require.Error(t, err)
+
+		// TODO: investigate why the message "failed to force-include tx" is not shown, it's a bug for OP, nothing to do with us
+
+	})
+
+	// tx.cost > acc.balance + acc.sgt_balance, should fail
+	t.Run("cost exceed native+sgt balance", func(t *testing.T) {
+		op_e2e.InitParallel(t)
+		cfg := e2esys.DefaultSystemConfigForSoulGasToken(t, true, true)
+
+		ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+		defer cancel()
+
+		opGeth, err := NewOpGeth(t, ctx, &cfg)
+		require.NoError(t, err)
+		defer opGeth.Close()
+
+		// mint enough SoulGasToken with deposit tx, but no native balance
+		targetPK, _ := crypto.GenerateKey()
+		targetAccount := crypto.PubkeyToAddress(*targetPK.Public().(*ecdsa.PublicKey))
+
+		// fund targetAccount with 2 wei
+		transferEthTx := types.NewTx(&types.DepositTx{
+			From:  cfg.Secrets.Addresses().Bob,
+			To:    &targetAccount,
+			Value: big.NewInt(params.Wei),
+			Gas:   1000001,
+		})
+		// convert 1 wei to sgt for targetAccount
+		mintTx := types.NewTx(&types.DepositTx{
+			From:  targetAccount,
+			To:    &types.SoulGasTokenAddr,
+			Value: big.NewInt(params.Wei),
+			Gas:   1000001,
+			Data:  SoulGasTokenABI.Methods["deposit"].ID,
+		})
+		_, err = opGeth.AddL2Block(ctx, transferEthTx, mintTx)
+		require.NoError(t, err)
+
+		receipt, err := opGeth.L2Client.TransactionReceipt(ctx, mintTx.Hash())
+		require.NoError(t, err)
+		assert.Equal(t, types.ReceiptStatusSuccessful, receipt.Status)
+
+		signer := types.LatestSigner(opGeth.L2ChainConfig)
+
+		tx := types.MustSignNewTx(targetPK, signer, &types.DynamicFeeTx{
+			ChainID:   big.NewInt(int64(cfg.DeployConfig.L2ChainID)),
+			Nonce:     1,
+			GasTipCap: big.NewInt(100),
+			GasFeeCap: big.NewInt(100000),
+			Gas:       1_000_001,
+			To:        &cfg.Secrets.Addresses().Alice,
+			Value:     big.NewInt(params.Wei),
+			Data:      nil,
+		})
+		_, err = opGeth.AddL2Block(ctx, tx)
+		require.Error(t, err)
+	})
+
+	// tx.value <= acc.balance && tx.cost <= acc.balance + acc.sgt_balance, should succeed
+	t.Run("happy path", func(t *testing.T) {
```

Comment:
Consider adding a boundary test case to check for depletion of both sgt and account balances. You might set `tx.value` to `acc.balance` and `tx.cost` to `acc.sgt_balance`

Reply:
It will fail, since tx.cost doesn't include l1 cost.